File: mlx-sys/src/linalg.rs
use cxx::{CxxVector, UniquePtr};

use crate::Optional;

type OptionalAxis = Optional<UniquePtr<CxxVector<i32>>>;

unsafe impl cxx::ExternType for OptionalAxis {
    type Id = cxx::type_id!("mlx_cxx::OptionalAxis");

    type Kind = cxx::kind::Opaque;
}

#[cxx::bridge]
pub mod ffi {
    unsafe extern "C++" {
        include!("mlx-cxx/linalg.hpp");

        #[namespace = "mlx::core"]
        type array = crate::array::ffi::array;

        #[namespace = "mlx_cxx"]
        type OptionalAxis = crate::linalg::OptionalAxis;

        #[namespace = "mlx_cxx"]
        type StreamOrDevice = crate::utils::StreamOrDevice;

        #[namespace = "mlx_cxx"]
        fn norm_ord(
            a: &array,
            ord: f64,
            axis: &OptionalAxis,
            keepdims: bool,
            s: StreamOrDevice,
        ) -> Result<UniquePtr<array>>;

        #[namespace = "mlx_cxx"]
        fn norm_ord_axis(
            a: &array,
            ord: f64,
            axis: i32,
            keepdims: bool,
            s: StreamOrDevice,
        ) -> Result<UniquePtr<array>>;

        #[namespace = "mlx_cxx"]
        fn norm_str_ord(
            a: &array,
            ord: &CxxString,
            axis: &OptionalAxis,
            keepdims: bool,
            s: StreamOrDevice,
        ) -> Result<UniquePtr<array>>;

        // #[namespace = "mlx_cxx"]
        // fn norm_str_ord_axis(
        //     a: &array,
        //     ord: &CxxString,
        //     axis: i32,
        //     keepdims: bool,
        //     s: StreamOrDevice,
        // ) -> Result<UniquePtr<array>>;

        #[namespace = "mlx_cxx"]
        fn norm(
            a: &array,
            axis: &OptionalAxis,
            keepdims: bool,
            s: StreamOrDevice,
        ) -> Result<UniquePtr<array>>;

        #[namespace = "mlx_cxx"]
        fn norm_axis(
            a: &array,
            axis: i32,
            keepdims: bool,
            s: StreamOrDevice,
        ) -> Result<UniquePtr<array>>;

        #[namespace = "mlx_cxx"]
        fn svd(a: &array, s: StreamOrDevice) -> Result<UniquePtr<CxxVector<array>>>;
    }
}

File: mlx-sys/src/linalg.cpp
#include "mlx/linalg.h"

#include "mlx-cxx/mlx_cxx.hpp"
#include "mlx-cxx/linalg.hpp"

namespace mlx_cxx {
    std::optional<std::vector<int>> to_std_optional(const OptionalAxis &opt) {
        switch (opt.tag) {
            case OptionalAxis::Tag::None:
                return std::nullopt;
            case OptionalAxis::Tag::Some:
                return *opt.payload.some;
        }
    }

    std::unique_ptr<mlx::core::array> norm_ord(
        const mlx::core::array &a,
        const double ord,
        const OptionalAxis &axis,
        bool keepdims,
        mlx_cxx::StreamOrDevice s)
    {
        auto axis_std = to_std_optional(axis);
        auto array = mlx::core::linalg::norm(a, ord, axis_std, keepdims, s.to_variant());
        return std::make_unique<mlx::core::array>(array);
    }

    std::unique_ptr<mlx::core::array> norm_ord_axis(
        const mlx::core::array& a,
        const double ord,
        int axis,
        bool keepdims,
        mlx_cxx::StreamOrDevice s)
    {
        auto array = mlx::core::linalg::norm(a, ord, axis, keepdims, s.to_variant());
        return std::make_unique<mlx::core::array>(array);
    }

    std::unique_ptr<mlx::core::array> norm_str_ord(
        const mlx::core::array& a,
        const std::string& ord,
        const OptionalAxis &axis,
        bool keepdims,
        mlx_cxx::StreamOrDevice s)
    {
        auto axis_std = to_std_optional(axis);
        auto array = mlx::core::linalg::norm(a, ord, axis_std, keepdims, s.to_variant());
        return std::make_unique<mlx::core::array>(array);
    }

    // std::unique_ptr<mlx::core::array> norm_str_ord_axis(
    //     const mlx::core::array& a,
    //     const std::string& ord,
    //     int axis,
    //     bool keepdims,
    //     mlx_cxx::StreamOrDevice s)
    // {
    //     auto array = mlx::core::linalg::norm(a, ord, axis, keepdims, s.to_variant());
    //     return std::make_unique<mlx::core::array>(array);
    // }

    std::unique_ptr<mlx::core::array> norm(
        const mlx::core::array& a,
        const OptionalAxis &axis,
        bool keepdims,
        mlx_cxx::StreamOrDevice s)
    {
        auto axis_std = to_std_optional(axis);
        auto array = mlx::core::linalg::norm(a, axis_std, keepdims, s.to_variant());
        return std::make_unique<mlx::core::array>(array);
    }

    std::unique_ptr<mlx::core::array> norm_axis(
        const mlx::core::array& a,
        int axis,
        bool keepdims,
        mlx_cxx::StreamOrDevice s)
    {
        auto array = mlx::core::linalg::norm(a, axis, keepdims, s.to_variant());
        return std::make_unique<mlx::core::array>(array);
    }

    std::array<std::unique_ptr<mlx::core::array>, 2> qr(
        const mlx::core::array& a,
        mlx_cxx::StreamOrDevice s)
    {
        auto [q, r] = mlx::core::linalg::qr(a, s.to_variant());
        return {std::make_unique<mlx::core::array>(q), std::make_unique<mlx::core::array>(r)};
    }
}